docs/articles/concepts/joins.md

```diff
@@ -1,3 +1,775 @@
+        "name": "Hive",
+    },
+]
+names = op.input("names", flow, TestingSource(names_l))
+
+emails_l = [
+    {
+        "user_id": 123,
+        "at": datetime(2023, 12, 14, 0, 15, tzinfo=timezone.utc),
+        "email": "bee@bytewax.io",
+    },
+    {
+        "user_id": 456,
+        "at": datetime(2023, 12, 14, 1, 15, tzinfo=timezone.utc),
+        "email": "hive@bytewax.io",
+    },
+]
+emails = op.input("emails", flow, TestingSource(emails_l))
+
+keyed_names = op.map("key_names", names, lambda x: (str(x["user_id"]), x))
+keyed_emails = op.map("key_emails", emails, lambda x: (str(x["user_id"]), x))
+
+joined = op_w.join_window("join", clock, windower, keyed_names, keyed_emails)
+
+op.inspect("check_join", joined)
+
+bytewax.testing.run_main(flow)
+```
+
+Looks like that's what we see! Notice the `None` in the output for key
+`"456"`. All window operators also emit the metadata about the window
+for analysis, but you can ignore that.
+
+```{testoutput}
+join_eg.check_join: ('456', (WindowMetadata(open_time: 2023-12-14 00:00:00 UTC, close_time: 2023-12-14 01:00:00 UTC), ({'user_id': 456, 'at': datetime.datetime(2023, 12, 14, 0, 0, tzinfo=datetime.timezone.utc), 'name': 'Hive'}, None)))
+join_eg.check_join: ('123', (WindowMetadata(open_time: 2023-12-14 00:00:00 UTC, close_time: 2023-12-14 01:00:00 UTC), ({'user_id': 123, 'at': datetime.datetime(2023, 12, 14, 0, 0, tzinfo=datetime.timezone.utc), 'name': 'Bee'}, {'user_id': 123, 'at': datetime.datetime(2023, 12, 14, 0, 15, tzinfo=datetime.timezone.utc), 'email': 'bee@bytewax.io'})))
+join_eg.check_join: ('456', (WindowMetadata(open_time: 2023-12-14 01:00:00 UTC, close_time: 2023-12-14 02:00:00 UTC), (None, {'user_id': 456, 'at': datetime.datetime(2023, 12, 14, 1, 15, tzinfo=datetime.timezone.utc), 'email': 'hive@bytewax.io'})))
+```
+
+Bytewax currently does not support running windowed joins.
+
+### Product Joins
+
+Window operators, because they have a defined close time, also support
+another join type. The **product join** emits all of the combinations
+of _all_ of the input values seen on a side.
+
+For example, if we don't change the join parameters, but update the
+input in the above dataflow to include multiple values in a window for
+a key.
+
+```python
+names_l.clear()
+names_l.extend(
+    [
+        {
+            "user_id": 123,
+            "at": datetime(2023, 12, 14, 0, 0, tzinfo=timezone.utc),
+            "name": "Bee",
+        },
+    ]
+)
+
+emails_l.clear()
+emails_l.extend(
+    [
+        {
+            "user_id": 123,
+            "at": datetime(2023, 12, 14, 0, 15, tzinfo=timezone.utc),
+            "email": "bee@bytewax.io",
+        },
+        {
+            "user_id": 123,
+            "at": datetime(2023, 12, 14, 0, 30, tzinfo=timezone.utc),
+            "email": "queen@bytewax.io",
+        },
+    ]
+)
+
+bytewax.testing.run_main(flow)
+```
+
+Notice how now we only have the latest email for the bee:
```

Comment:
to clarify, this example is great and fits the same user-id reasons, and let's keep it. But everywhere else we probably want to change that